### ISSUE TITLE
Performance drops while changing the order of dfl channels

### ISSUE THREAD

**Issue by wcyjerry** [ultralytics/ultralytics#3730]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

Hi, I was using dfl, and I need to use it on device, the device don't support transpose or permute op but support review op
So, during the inference, I change the DFL forward function
return self.conv(x.view(b, 4, self.c1, a).transpose(2, 1).softmax(1)).view(b, 4, a)  -> return self.conv(x.view(b, self.c1, 4, a).softmax(1)).view(b, 4, a)

and in the yolov8_loss, I changed the decode box function to make sure the order is aligned
pred_dist = pred_dist.view(b, a, 4, c // 4).softmax(3).matmul(self.proj.type(pred_dist.dtype)) -> pred_dist = pred_dist.view(b, a, c // 4, 4).transpose(2,3).softmax(3).matmul(self.proj.type(pred_dist.dtype))

This seems alright, while the results drops like 2~3 point in ap50 and ap50:95 in my task?
Is there something I missed to change in the mean time?

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@wcyjerry hi there! Thank you for reaching out and explaining your issue with YOLOv8.

Based on your description, it seems like you made some modifications to the DFL forward function and the decode box function in order to align the order of DFL channels with the capabilities of your device. However, you noticed a drop in performance, specifically in the ap50 and ap50:95 metrics.

Without reviewing the modified code, it's difficult to pinpoint the exact cause of the performance drop. However, there are a few potential reasons that could contribute to this issue:

1. Data compatibility: It's possible that the modified code is not correctly handling the input and output data, leading to a mismatch in format or information loss. Make sure that the changes you made are compatible with the data requirements of YOLOv8.

2. Mathematical operations: The modified code may introduce unintended changes to the mathematical operations performed during inference or decoding. Double-check that your changes do not alter the underlying calculations in a way that negatively impacts the performance.

3. Model assumptions: YOLOv8 has certain assumptions and design choices that contribute to its overall performance. Modifying the code may violate some of these assumptions, leading to unexpected behavior. Consider whether the changes you made are consistent with the underlying principles of YOLOv8.

To further diagnose the issue, I would recommend reviewing the code modifications in detail and comparing them with the original implementation. Additionally, you can validate your changes with a minimal reproducible example or by comparing the results against the original YOLOv8 to ensure their correctness.

If you believe that your changes are valid and can improve the performance of YOLOv8, you are more than welcome to submit a pull request (PR) so that the Ultralytics team and the YOLO community can review and potentially integrate your contributions.

I hope this information helps you in resolving your issue. If you have any further questions or need additional assistance, please don't hesitate to ask.

**Comment by wcyjerry** [comment]
@glenn-jocher Thank you for your detailed response. Even though this task originates from a personal project of yours, I appreciate your willingness to investigate the issue. 
I realize this is a particularly tricky question, primarily revolving around the fact that reshaping the 64 channels from the 'reg' branch into 16x4 and reshaping into 4x16 seem to yield different results. This is intriguing since theoretically, both should be equivalent provided the loss calculation is adjusted correspondingly. By the way, I'm using your official code, in which the default partition is 4x16. Yet, in the DFL code, a commented line indicates a 16x4 partition, and likewise, there's a commented out reshape into 16x4 in 'v8DetectionLoss'. It appears that both partitioning methods were considered, but you defaulted to the former.
![image](https://github.com/ultralytics/ultralytics/assets/48686306/a8b1f9f5-7642-46fe-ab1f-fd31034a1f29)
![image](https://github.com/ultralytics/ultralytics/assets/48686306/95351e3b-683b-4459-bd0c-93651f343dfb)
The default results in 0.81 of ap50, while the commented one only 0.78!

**Comment by wcyjerry** [comment]
> @glenn-jocher @wcyjerry Do you have any idea about why the default code achieve higher accuracy than the commented version? I am curious about this issue because if we can use commented version then we can save a transpose operation during inference.

Sry, I still don't figure out why this would happen, maybe it's about pre-trained model trained with (b, 4, 16, a). I will check trained from scratch with two different version.
As for "I am curious about this issue because if we can use commented version then we can save a transpose operation during inference." That's exactly why I'm trying to do this.

**Comment by wcyjerry** [comment]
@ZhixiongSun It will be grateful if you can verify whether the commented code will drop ap on your dataset. I'm curious that if it just happened on my case.

**Comment by ZhixiongSun** [comment]
@wcyjerry Hi,  Yes I have tried it on my dataset and the accuracy dropped if I used commented version code. It seems like the transpose module is necessary.

**Comment by wcyjerry** [comment]
While it doesnt make any sense since there completely equal&nbsp;





------------------ Original ------------------
From: ZhixiongSun ***@***.***&gt;
Date: Mon,Aug 14,2023 7:26 PM
To: ultralytics/ultralytics ***@***.***&gt;
Cc: Chenyi Wang ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [ultralytics/ultralytics] Performance drops while changing theorder of dfl channels (Issue #3730)





 
@wcyjerry Hi,  Yes I have tried it on my dataset and the accuracy dropped if I used commented version code. It seems like the transpose module is necessary.

**Comment by wcyjerry** [comment]
@glenn-jocher Could you plz check this out? or I just wonder why you guys chose the channel in the order of (b,4,16,a), cause the commented code indicated you tried another order?

**Comment by wcyjerry** [comment]
@ZhixiongSun  Hi, could you plz run two more experiments, using two version of codes without pretrained model.
See if the pretrained model will effect the result with different order of dfl channels.
